### DISCUSSION TITLE
(reverse order) Sort of especially in-addr.arpa zones/records

### DISCUSSION THREAD

**Discussion by hvisage** [poweradmin/poweradmin#425] [Ideas]
Let's look at this: the order seem.. odd when you are looking for a specific  zone/record, though, it's technically correct from a strings ordering perspective

![image](https://user-images.githubusercontent.com/11662665/159134975-4802205d-8da5-46a3-8239-c46825cdd76b.png)

Below are some PostgreSQL functions, that provides some beter UI sort order and views (at least for me):

![image](https://user-images.githubusercontent.com/11662665/159135219-e8161927-7d2c-4a09-94ee-64a554fa2238.png)

I know @edmondas is busy with active refactoring, so just wondering where/how I could assist n adding these?

```
--
-- Name: hnum(character varying); Type: FUNCTION
--

CREATE FUNCTION public.hnum(hstr character varying) RETURNS character varying
    LANGUAGE plpgsql IMMUTABLE
    AS $$
begin
return trim(to_char(hstr::numeric,'000'));
exception
 when others then
 return hstr;
end;
$$;

-- Name: reverse_0fqdn(character varying); Type: FUNCTION; Schema: public; Owner: postgres

CREATE FUNCTION public.reverse_0fqdn(fname character varying) RETURNS character varying
    LANGUAGE plpgsql IMMUTABLE
    AS $$
declare
 rname varchar[];
 x varchar;
 z numeric(3);
BEGIN
foreach x in array string_to_array(fname,'.')
loop
 rname = array_prepend(hnum(x),rname);
end loop;
return array_to_string(rname,'.');
end;
$$;

-- Name: reverse_fqdn(character varying); Type: FUNCTION; Schema: public; Owner: postgres

CREATE FUNCTION public.reverse_fqdn(fname character varying) RETURNS character varying
    LANGUAGE plpgsql IMMUTABLE
    AS $$
declare
 rname varchar[];
 x varchar;
BEGIN
foreach x in array string_to_array(fname,'.')
loop
 rname = array_prepend(x,rname);
end loop;
return array_to_string(rname,'.');
end;
$$;

```

**Comment by hvisage** [comment]
![image](https://user-images.githubusercontent.com/11662665/159135357-9b7ce58f-72db-4216-95e9-d64e6e6e9847.png)

**Comment by edmondas** [comment]
What do you think if it is sorted like this:
![Screenshot 2022-03-20 at 15 14 38](https://user-images.githubusercontent.com/9033112/159164076-99e30103-9ee3-42e7-ad94-8609618b3124.png)

It only requires a few lines of code.

**Reply by edmondas** [reply]
Initially, filter out reverse domains, then sort by domain length that includes numbers, and finally sort by regular name:

```diff
diff --git a/lib/DnsRecord.php b/lib/DnsRecord.php
index babd919..2f69abd 100644
--- a/lib/DnsRecord.php
+++ b/lib/DnsRecord.php
@@ -1134,6 +1134,8 @@ class DnsRecord
         $natural_sort = 'domains.name';
         if ($db_type == 'mysql' || $db_type == 'mysqli' || $db_type == 'sqlite' || $db_type == 'sqlite3') {
             $natural_sort = 'domains.name+0<>0 DESC, domains.name+0, domains.name';
+        } elseif ($db_type == 'pgsql') {
+            $natural_sort = "SUBSTRING(domains.name FROM '\.arpa$'), LENGTH(SUBSTRING(domains.name FROM '^[0-9]+')), domains.name";
         }
         $sql_sortby = ($sortby == 'domains.name' ? $natural_sort : $sortby . ', ' . $natural_sort);
 
```

**Reply by hvisage** [reply]
> 

Where would 251.252.10 sort w.r.t. 1.16.172, 128.32.172 and 2.252.10 ?

The ideal with my sorting is you do sort :
10
1.10
252.1.10
2.10
100.100.10
16.172
200.1.168.192
1.2.168.192
2.255.168.192

easier to get to the right domain

also, another sorting that it will solve:

example.com
finances.example.com
zero.example.com
test.com
axe.test.com
example.net
test.net

when you are looking for a subdomain - but then, I guess it's a personal search/sort preference

**Reply by hvisage** [reply]
Which then starts to be something like going down the domains in levels (except the 2 and 3 levels TLDs might be a "fun exceptions"),

ie. only show levels 1, end then clicking on that (if there are extra levels) show them as zones, on top (and then the zone's details below) like directories & files in file finders

**Reply by edmondas** [reply]
A domain list expansion requires the introduction of an API and a javascript framework in order to simplify the implementation. With the current code base, it's a bit too early to implement such features.

**Reply by edmondas** [reply]
Another option would be to separate the lists (or pages):
1st - for reverse zones
2nd - for regular zones with expandable blocks (for quick access to 2nd/3rd level domains)

**Comment by hvisage** [comment]
With the 3.x, is the Javascript framework still the challenge for this type of sorting?

**Reply by edmondas** [reply]
> With the 3.x, is the Javascript framework still the challenge for this type of sorting?

Perhaps the issue is not with the JS framework but rather with the lack of APIs that provide data.

**Comment by edmondas** [comment]
Please check implementation ticket #455 for things that I've done for next release. Closing this discussion now.
